lib/storage/data/file/store.js

```diff
@@ -0,0 +1,274 @@
+                                   (v, k) => (k).toString());
+        this.logger.info(`pre-creating ${subDirs.length} subdirs...`);
+        if (!this.noSync) {
+            storageUtils.setDirSyncFlag(this.dataPath, this.logger);
+        }
+        async.eachSeries(subDirs, (subDirName, next) => {
+            fs.mkdir(`${this.dataPath}/${subDirName}`, err => {
+                // If already exists, move on
+                if (err && err.code !== 'EEXIST') {
+                    return next(err);
+                }
+                return next();
+            });
+        },
+        err => {
+            assert.strictEqual(err, null, `Error creating data files ${err}`);
+            this.logger.info('data file store init complete, ' +
+                             'go forth and store data.');
+            callback();
+        });
+    }
+
+    /**
+     * Get the filesystem path to a stored object file from its key
+     *
+     * @param {String} key - the object key
+     * @return {String} the absolute path to the file containing the
+     *   object contents
+     */
+    getFilePath(key) {
+        const hash = stringHash(key);
+        const folderHashPath = ((hash % FOLDER_HASH)).toString();
+        return `${this.dataPath}/${folderHashPath}/${key}`;
+    }
+
+    /**
+     * Put a new object to the storage backend
+     *
+     * @param {stream.Readable} dataStream - input stream providing the
+     *   object data
+     * @param {Number} size - Total byte size of the data to put
+     * @param {Logger} log - logging object
+     * @param {DataFileStore~putCallback} callback - called when done
+     * @return {undefined}
+     */
+    put(dataStream, size, log, callback) {
+        const key = crypto.pseudoRandomBytes(20).toString('hex');
+        const filePath = this.getFilePath(key);
+        log.debug('starting to write data', { key, filePath });
+        dataStream.pause();
+        fs.open(filePath, 'wx', (err, fd) => {
+            if (err) {
+                log.error('error opening filePath', { error: err });
+                return callback(err);
+            }
+            const cbOnce = jsutil.once(callback);
+            // disable autoClose so that we can close(fd) only after
+            // fsync() has been called
+            const fileStream = fs.createWriteStream(filePath,
+                                                    { fd,
+                                                      autoClose: false });
+
+            fileStream.on('finish', () => {
+                function ok() {
+                    log.debug('finished writing data', { key });
+                    return cbOnce(null, key);
+                }
+                if (this.noSync) {
+                    fs.close(fd);
+                    return ok();
+                }
+                fs.fsync(fd, err => {
+                    fs.close(fd);
+                    if (err) {
+                        log.error('fsync error', { error: err });
```

Comment:
Please annotate the logs with `method` param so that it's easier when debugging.